sbi/samplers/mcmc/pymc_utils.py

```diff
@@ -0,0 +1,173 @@
+        """Get gradients computed from `perform` and return Jacobian-Vector product"""
+        # get outputs from forward pass (but doesn't re-compute it, I think...)
+        value = self(*inputs)
+        gradients = value.owner.outputs[1:]
+        # compute and return JVP
+        return [(g[0] * grad) for grad in gradients]
+
+
+class PyMCSampler:
+    def __init__(
+        self,
+        potential_fn: Callable,
+        initvals: np.ndarray,
+        step: str = "nuts",
+        draws: int = 1000,
+        tune: int = 1000,
+        chains: Optional[int] = None,
+        mp_ctx: str = "spawn",
+        progressbar: bool = True,
+        param_name: str = "theta",
+        device: str = "cpu",
+    ):
+        """Interface for PyMC samplers
+
+        Args:
+            potential_fn: Potential function from density estimator.
+            initvals: Initial parameters.
+            step: One of `"slice"`, `"hmc"`, or `"nuts"`.
+            draws: Number of total samples to draw.
+            tune: Number of tuning steps to take.
+            chains: Number of MCMC chains to run in parallel.
+            mp_ctx: Multiprocessing context for parallel sampling.
+            progressbar: Whether to show/hide progress bars.
+            param_name: Name for parameter variable, for PyMC and ArviZ structures
+            device: The device to which to move the parameters for potential_fn.
+        """
+        self.param_name = param_name
+        self._step = step
+        self._draws = draws
+        self._tune = tune
+        self._initvals = [{self.param_name: iv} for iv in initvals]
+        self._chains = chains
+        self._mp_ctx = mp_ctx
+        self._progressbar = progressbar
+        self._device = device
+
+        # create PyMC model object
+        track_gradients = True if step in (pymc.NUTS, pymc.HamiltonianMC) else False
+        self._model = pymc.Model()
+        potential = PyMCPotential(
+            potential_fn, track_gradients=track_gradients, device=device
+        )
+        with self._model:
+            params = pymc.Normal(
+                self.param_name, mu=initvals.mean(axis=0)
+            )  # dummy prior
+            pymc.Potential("likelihood", potential(params))
+
+    def run(self) -> np.ndarray:
+        """Run MCMC with PyMC
+
+        Returns:
+            MCMC samples
+        """
+        step_class = dict(slice=pymc.Slice, hmc=pymc.HamiltonianMC, nuts=pymc.NUTS)
+        with self._model:
+            inference_data = pymc.sample(
+                step=step_class[self._step](),
+                tune=self._tune,
+                draws=self._draws,
+                initvals=self._initvals,
+                chains=self._chains,
+                progressbar=self._progressbar,
+                mp_ctx=self._mp_ctx,
+            )
+        self._inference_data = inference_data
+        samples = getattr(inference_data.posterior, self.param_name).data
+        return samples
+
+    def get_samples(
```

Comment:
what was your reasoning for having the methods `run` and `get_samples` (separately)? Is there an API interface in sbi for samplers?

Reply:
the `run()` method is not necessary and just maintains consistency with other samplers, but the `get_samples()` function is required for `MCMCPosterior.get_arviz_inference_data` to work. This is not explicitly documented anywhere, but seems like the recommended Python approach here might be to define a (Protocol)[https://peps.python.org/pep-0544/]. perhaps there could be some discussion on whether this is necessary